src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3808 @@
+    println(io, "defined on the Singular side")
+  end
+end
+
+function length(M::SubModuleOfFreeModule)
+  #TODO replace all calls by ngens
+  error("use ngens() instead of length()")
+  return length(M.gens)
+end
+
+@doc Markdown.doc"""
+  ngens(M::SubModuleOfFreeModule)
+
+Return the number of generators of `M`.
+"""
+function ngens(M::SubModuleOfFreeModule)
+  return ngens(M.gens)
+end
+
+@doc Markdown.doc"""
+  gens(M::SubModuleOfFreeModule)
+
+Return the generators of `M` as an array of `FreeModElem`s.
+"""
+function gens(M::SubModuleOfFreeModule)
+  return oscar_generators(M.gens)
+end
+
+@doc Markdown.doc"""
+  gen(M::SubModuleOfFreeModule, i::Int)
+
+Return the `i`th generator of `M`.
+"""
+function gen(M::SubModuleOfFreeModule, i::Int)
+  return oscar_generators(M.gens)[i]
+  #return M.gens[Val(:O), i]
+end
+
+@doc Markdown.doc"""
+  sum(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+
+Compute $M+N$.
+"""
+function Base.sum(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+  @assert M.F === N.F
+  return SubModuleOfFreeModule(M.F, vcat(collect(M.gens), collect(N.gens)))
+end
+
+@doc Markdown.doc"""
+  issubset(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+
+Check if `M` is a subset of `N`. For this their embedding free modules must be 
+identical (`===`).
+"""
+function Base.issubset(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+  @assert M.F === N.F
+  M_mod_N = _reduce(singular_generators(std_basis(M)), singular_generators(std_basis(N)))
+  return iszero(M_mod_N)
+end
+
+@doc Markdown.doc"""
+  ==(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+
+Check for equality. For two submodules of free modules to be equal their embedding 
+free modules must be identical (`===`) and the generators must generate equal submodules.
+"""
+function Base.:(==)(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule)
+  @assert M.F === N.F
+  M_mod_N = _reduce(singular_generators(std_basis(M)), singular_generators(std_basis(N)))
+  N_mod_M = _reduce(singular_generators(std_basis(N)), singular_generators(std_basis(M)))
+  return iszero(M_mod_N) && iszero(N_mod_M)
+end
+
+#+(M::SubModuleOfFreeModule, N::SubModuleOfFreeModule) = sum(M, N)
+
+@doc Markdown.doc"""
+  SubQuo{T} <: ModuleFP{T}
+
+The type of subquotient modules.
+A subquotient module $M$ is a module where $M = A + B / B$ where $A$ and $B$ are 
+submodules of a free module.
+$A$, $B$ and $A+B$ (they have type `SubModuleOfFreeModule`) as well as the embedding
+free module are stored. 
+One can construct ordinary submodules of free modules by not giving $B$.
+Moreover, canonical in- and outgoing morphisms are stored if the corresponding
+option is set in suitable functions.
+`SubQuo{T}` is a subtype of `ModuleFP{T}`.
+"""
+mutable struct SubQuo{T} <: ModuleFP{T}
+  #meant to represent sub+ quo mod quo - as lazy as possible
+  F::FreeMod{T}
+  sub::SubModuleOfFreeModule
+  quo::SubModuleOfFreeModule
+  sum::SubModuleOfFreeModule
+
+  ingoing_morphisms::Vector{<:ModuleMap}
```

Comment:
I don't think "ingoing" is a proper english word. Perhaps "incoming"
```suggestion
  incoming_morphisms::Vector{<:ModuleMap}
```